src/server.rs

```diff
@@ -1,62 +1,119 @@
-//     B: Write,
-// {
-//     let status = rsp.status();
-//     let headers = rsp.headers().clone(); // TODO: remove clone.
-//     let body: B = *(rsp.body_ref().unwrap().clone());
-//     let body = f(body);
-//     http::Response::new()
-//         .with_status(status)
-//         .with_headers(headers)
-//         .with_body(body)
-// }
-
-impl<T> Gzip<T> {
+const MAX_CHUNK_SIZE: usize = 2048;
+
+struct ChunkingStream<B: AsyncRead> {
+    read: B,
+    chunksz: usize,
+}
+
+// impl Builder
+impl Builder {
+    pub fn chunk_size(mut self, size: usize) -> Self {
+        self.chunk_size = Some(size);
+        self
+    }
+
+    pub fn compression_level(mut self, compression: Compression) -> Self {
+        self.compression = Some(compression);
+        self
+    }
+
+    pub fn to_service<T>(&self, inner: T) -> GzipChunked<T> {
+        GzipChunked {
+            inner,
+            chunk_size: self.chunk_size.unwrap_or(MAX_CHUNK_SIZE),
+            compression: self.compression.unwrap_or_default(),
+        }
+    }
+}
+
+impl<B: AsyncRead> Stream for ChunkingStream<B> {
+    type Item = Bytes;
+    type Error = hyper::Error;
+
+     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let mut buf = vec![0u8; self.chunksz];
+        let read = try_ready!(
+            self.read
+                .poll_read(&mut buf)
+                .map_err(hyper::Error::Io)
+        );
+        trace!("ChunkingStream: read {:?} bytes", read);
+
+        if read == 0 {
+            trace!("ChunkingStream: reader emptied");
+            return Ok(Async::Ready(None));
+        }
+
+        let chunk = Bytes::from(&buf[..read]);
+        trace!("ChunkingStream: wrote {:?} byte chunk", chunk.len());
+        Ok(Async::Ready(Some(chunk)))
+
+
+     }
+}
+impl GzipChunked<()> {
+
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+}
+
+impl<T> GzipChunked<T> {
+
     pub fn new(inner: T) -> Self {
-        Self { inner, }
+        Builder::default().to_service(inner)
     }
+
 }
 
 fn is_gzip<A>(req: &http::Request<A>) -> bool {
-    // let accept_encodings = req.headers()
-    //     .get::<header::AcceptEncoding>();
-    // if accept_encodings.is_none() {
-    //     return false;
-    // }
-    true
+    if let Some(accept_encodings) = req
+        .headers()
+        .get(header::ACCEPT_ENCODING)
+    {
+        // TODO: honor quality items & stuff.
+        return accept_encodings
+            .to_str()
+            .expect("accept-encoding header shouldn't be a weird wad of binary")
+            .contains("gzip");
+    }
+    false
 }
 
-impl<T, A, B> Service for Gzip<T>
+impl<T, A, B> Service for GzipChunked<T>
 where
     T: Service<
         Request = http::Request<A>,
         Response = http::Response<B>,
         Error = hyper::Error,
     >,
     T::Future: Send + 'static,
-    B: Into<Body> + AsRef<[u8]>,
-    // MaybeCompressed<B>: AsyncRead,
+    B: Into<Body> + AsRef<[u8]> + Send + 'static,
```

Comment:
Is the goal to wrap only bodies that have an `AsRef<[u8]>`? Or would you also want to encode any stream that doesn't have the full body loaded in memory yet?